py3status/modules/velib_metropole.py

```diff
@@ -0,0 +1,292 @@
+        self.gps_bot_left_latitude = 48.6
+        self.gps_bot_left_longitude = 2.2
+        self.gps_zoom_level = 15
+
+        self.station_states = {"Operative": 0, "Close": 1}
+
+        json = self._get_velib_data()
+        if json:
+            self.number_of_stations, self.stations = self._manipulate(json)
+            self._get_usefull_area(self.stations)
+
+        # get placeholders
+        self.placeholders = []
+        for x in ["format", "format_station"]:
+            self.placeholders.append(self.py3.get_placeholders_list(x))
+
+        self.refresh = True
+        self.station_index = 1
+
+    def _camel_to_snake_case(self, data):
+        if not isinstance(data, (int, float)):
+            s1 = sub("(.)([A-Z][a-z]+)", r"\1_\2", data)
+            return sub("([a-z0-9])([A-Z])", r"\1_\2", s1).lower()
+        else:
+            return data
+
+    def _cast_number(self, value):
+        try:
+            value = float(value)
+        except ValueError:
+            try:
+                value = int(value)
+            except ValueError:
+                pass
+        return value
+
+    def _get_usefull_area(self, data):
+        """
+        reduce the zone to reduce the size of fetched data on refresh
+        """
+
+        latitudes = []
+        longitudes = []
+        for x in data:
+            latitudes.append(float(data[x]["station_gps_latitude"]))
+            longitudes.append(float(data[x]["station_gps_longitude"]))
+        self.gps_top_right_latitude = max(latitudes)
+        self.gps_top_right_longitude = max(longitudes)
+        self.gps_bot_left_latitude = min(latitudes)
+        self.gps_bot_left_longitude = min(longitudes)
+
+    def _get_velib_data(self):
+        url = "".join(
+            [
+                "https://www.velib-metropole.fr/webapi/map/details?",
+                "gpsTopLatitude=%s",
+                "&gpsTopLongitude=%s",
+                "&gpsBotLatitude=%s",
+                "&gpsBotLongitude=%s",
+                "&zoomLevel=%s",
+            ]
+        )
+        url = url % (
+            self.gps_top_right_latitude,
+            self.gps_top_right_longitude,
+            self.gps_bot_left_latitude,
+            self.gps_bot_left_longitude,
+            self.gps_zoom_level,
+        )
+
+        try:
+            return self.py3.request(url, timeout=self.request_timeout).json()
+        except self.py3.RequestException:
+            return None
+
+    def _manipulate(self, data):
+        index = 1
+        stations = {}
+
+        for code in self.station_codes:
+            new_station = {}
+
+            # search station
+            station = next(
+                (item for item in data if int(item["station"]["code"]) == int(code)),
+                None,
+            )
+
+            # flat, camel to snake and cast...
+            for k, v in self.py3.flatten_dict(station, delimiter="_").items():
+                new_station[self._camel_to_snake_case(k)] = self._cast_number(v)
```

Comment:
You're running this all times. This and `datetime.fromtimestamp` below too. Chance is that users don't always use them so you should see if you can be more efficient by only working with placeholders first instead of running this on all variables in every station.... for unrelated variables on only one station.